tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -247,119 +244,205 @@ def _external_dep_name_from_bazel_dependency(bazel_dep):
 
 
-def _expand_intermediate_deps(target_dict, public_dep_names, bazel_rules):
-    # Some of the libraries defined by bazel won't be exposed in build.yaml
-    # We call these "intermediate" dependencies. This method expands
-    # the intermediate deps for given target (populates library's
-    # headers, sources and dicts as if the intermediate dependency never existed)
-
-    # use this dictionary to translate from bazel labels to dep names
-    bazel_label_to_dep_name = {}
-    for dep_name in public_dep_names:
-        bazel_label_to_dep_name[_get_bazel_label(dep_name)] = dep_name
+def _deduplicate_append(target: List[str], pending: List[str]) -> None:
+    """Append the list without duplicated items in place.
+    
+    Expected to be used to create a order-stable dependency list.
+    """
+    seen = set(target)
+    for dep in pending:
+        if dep not in seen:
+            target.append(dep)
+
+
+def _compute_transitive_metadata(
+        rule_name: str, bazel_rules: Any,
+        bazel_label_to_dep_name: Dict[str, str]) -> None:
+    """Computes the final build metadata for Bazel target with rule_name.
+
+    The dependencies that will appear on the deps list are:
+
+    * Public build targets including binaries and tests;
+    * External targets, like absl, re2.
+
+    All other intermediate dependencies will be merged, which means their
+    source file, headers, etc. will be collected into one build target. This
+    step of processing will greatly reduce the complexity of the generated
+    build specifications for other build systems, like CMake, Make, setuptools.
+
+    The final build metadata are:
+    * _TRANSITIVE_DEPS: all the transitive dependencies including intermediate
+                        targets;
+    * _COLLAPSED_DEPS:  dependencies that fits our requirement above, and it
+                        will remove duplicated items and produce the shortest
+                        possible dependency list in alphabetical order;
+    * _COLLAPSED_SRCS:  the merged source files;
+    * _COLLAPSED_PUBLIC_HEADERS: the merged public headers;
+    * _COLLAPSED_HEADERS: the merged non-public headers;
+    * _EXCLUDE_DEPS: transitive dependencies of public deps that this build
+      target depends on, and they will be excluded in the final build metadata.
+
+    For the collapsed_deps, the algorithm improved cases like:
+
+    The result in the past:
+        end2end_tests -> [grpc_test_util, grpc, gpr, address_sorting, upb]
+        grpc_test_util -> [grpc, gpr, address_sorting, upb, ...]
+        grpc -> [gpr, address_sorting, upb, ...]
+    
+    The result of the algorithm:
+        end2end_tests -> [grpc_test_util]
+        grpc_test_util -> [grpc]
+        grpc -> [gpr, address_sorting, upb, ...]
+    """
+    bazel_rule = bazel_rules[rule_name]
+    direct_deps = _extract_deps(bazel_rule, bazel_rules)
+    transitive_deps = []
+    # We care about the order of dependencies list, so it will be a list.
```

Comment:
nit:  "We care about the order of dependencies list, so it will be a list. " is no longer true.

besides that, does `transitive_deps = []` really need to be a list?